torchdata/stateful_dataloader/sampler.py

```diff
@@ -125,3 +126,46 @@ def __iter__(self):
                     batch = [0] * self.batch_size
             if idx_in_batch > 0:
                 yield batch[:idx_in_batch]
+
+
+class StatefulDistributedSampler(torch.utils.data.distributed.DistributedSampler):
+    _YIELDED = "yielded"
+
+    def __init__(
+        self,
+        dataset: Dataset,
+        num_replicas: Optional[int] = None,
+        rank: Optional[int] = None,
+        shuffle: bool = True,
+        seed: int = 0,
+        drop_last: bool = False,
+    ) -> None:
+        super().__init__(dataset, num_replicas, rank, shuffle, seed, drop_last)
+        self.yielded = 0
+        self.next_yielded = None
+        self.current_index = 0
+
+    def __iter__(self):
+        self.indices = list(super().__iter__())
+        self.current_index = 0
+        return self
+
+    def __next__(self) -> int:
+        if self.next_yielded is not None:
+            self.current_index = self.next_yielded
+            self.yielded = self.next_yielded
+            self.next_yielded = None
+        if self.current_index >= len(self.indices):
+            raise StopIteration
+        val = self.indices[self.current_index]
+        self.current_index += 1
+        self.yielded += 1
+        return val
```

Comment:
This will make a copy of the whole index list. Since DistributedSampler has a copy of the whole list anyways, the most efficient would be to randomly index into that, but we'd need to expose it or copy/paste that code into here. Instead let's just use itertools slice to avoid making the second copy of the list

```suggestion
    def __iter__(self):
        self.yielded = 0
        if self.next_yielded is not None:
            self.yielded = self.next_yielded
            self.next_yielded = None
        it = super().__iter__()
        for idx in itertools.islice(it, self.yielded, None):
            self.yielded += 1
            yield idx
```